linux/libcmt/include/libcmt/rollup.h

```diff
@@ -0,0 +1,168 @@
+
+/** Public struct with the advance state contents */
+typedef struct {
+	uint8_t  sender[CMT_ADDRESS_LENGTH]; /**< the address of the input sender */
+	uint64_t block_number;               /**< block number of this input */
+	uint64_t block_timestamp;            /**< block timestamp of this input UNIX epoch format) */
+	uint64_t index;                      /**< input index (in relation to all inputs ever sent to the DApp) */
+	size_t   length;                     /**< length in bytes of the data field */
+	void    *data;                       /**< advance contents */
+} cmt_rollup_advance_t;
+
+/** Public struct with the inspect state contents */
+typedef struct {
+	size_t length; /**< length in bytes of the data field */
+	void  *data;   /**< inspect contents */
+} cmt_rollup_inspect_t;
+
+/** Public struct with the finish state contents */
+typedef struct {
+	bool accept_previous_request;
+	int next_request_type;
+	size_t next_request_payload_length;
+} cmt_rollup_finish_t;
+
+/** Initialize a @ref cmt_rollup_t state.
+ *
+ * @param [in] me    uninitialized state
+ *
+ * - 0 success
+ * - negative value on error. values from: @ref cmt_rollup_driver_init */
+int
+cmt_rollup_init(cmt_rollup_t *me);
+
+/** Finalize a @ref cmt_rollup_t statate previously initialized with @ref
+ * cmt_rollup_init
+ *
+ * @param [in] me    initialized state
+ *
+ * @note use of @p me after this call is undefined behavior. */
+void
+cmt_rollup_fini(cmt_rollup_t *me);
+
+/** Emit a voucher
+ *
+ * @param [in,out] me      initialized cmt_rollup_t instance
+ * @param [in]     address destination
+ * @param [in]     n       sizeof @p data in bytes
+ * @param [in]     data    message contents
+ * @return
+ * - 0 success
+ * - -ENOBUFS no space left in @p me */
+int
+cmt_rollup_emit_voucher(cmt_rollup_t *me
+                       ,uint8_t address[CMT_ADDRESS_LENGTH]
+                       ,size_t n
+                       ,const void *data);
+
+/** Emit a notice
+ *
+ * @param [in,out] me      initialized cmt_rollup_t instance
+ * @param [in]     n       sizeof @p data in bytes
+ * @param [in]     data    message contents
+ * @return
+ * - 0 success
+ * - -ENOBUFS no space left in @p me */
+int
+cmt_rollup_emit_notice(cmt_rollup_t *me
+                      ,size_t n
+                      ,const void *data);
+
+/** Emit a report
+ * @param [in,out] me      initialized cmt_rollup_t instance
+ * @param [in]     n       sizeof @p data in bytes
+ * @param [in]     data    message contents
+ * @return
+ * - 0 success
+ * - -ENOBUFS no space left in @p me */
+int
+cmt_rollup_emit_report(cmt_rollup_t *me
+                      ,size_t n
+                      ,const void *data);
+
+/** Emit a exception
+ * @param [in,out] me      initialized cmt_rollup_t instance
+ * @param [in]     n       sizeof @p data in bytes
+ * @param [in]     data    message contents
+ * @return
+ * - 0 success
+ * - -ENOBUFS no space left in @p me */
+int
+cmt_rollup_emit_report(cmt_rollup_t *me
```

Comment:
I think that the name of this function is wrong.

Reply:
What a terrible copy pasta... nice catch.